File: core/src/main/java/io/snabble/sdk/CheckoutApi.java
package io.snabble.sdk;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.annotations.SerializedName;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

import io.snabble.sdk.payment.PaymentCredentials;
import io.snabble.sdk.utils.DateUtils;
import io.snabble.sdk.utils.GsonHolder;
import io.snabble.sdk.utils.JsonCallback;
import io.snabble.sdk.utils.Logger;
import io.snabble.sdk.utils.SimpleJsonCallback;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class CheckoutApi {
    private static final MediaType JSON = MediaType.parse("application/json");

    /*
     * Data structures as defined here:
     *
     * https://github.com/snabble/docs/blob/master/api_checkout.md
     */
    public static class Href {
        public String href;
    }

    public static class SignedCheckoutInfo {
        public JsonObject checkoutInfo;
        public String signature;
        public Map<String, Href> links;

        public String getCheckoutProcessLink() {
            Href checkoutProcess = links.get("checkoutProcess");
            if (checkoutProcess != null && checkoutProcess.href != null) {
                return checkoutProcess.href;
            }
            return null;
        }

        public PaymentMethodInfo[] getAvailablePaymentMethods(PaymentMethod[] clientAcceptedPaymentMethods) {
            if (checkoutInfo != null && checkoutInfo.has("paymentMethods")) {
                JsonArray jsonArray = checkoutInfo.getAsJsonArray("paymentMethods");
                if (jsonArray != null) {
                    List<PaymentMethodInfo> paymentMethods = new Gson().fromJson(jsonArray,
                            new TypeToken<List<PaymentMethodInfo>>() {}.getType());

                    if (clientAcceptedPaymentMethods == null) {
                        clientAcceptedPaymentMethods = PaymentMethod.values();
                    }

                    List<PaymentMethodInfo> result = new ArrayList<>();

                    for (PaymentMethod clientPaymentMethod : clientAcceptedPaymentMethods) {
                        for (PaymentMethodInfo paymentMethodInfo : paymentMethods) {
                            PaymentMethod pm = PaymentMethod.fromString(paymentMethodInfo.id);
                            if (pm != null) {
                                if (pm.equals(clientPaymentMethod)) {
                                    result.add(paymentMethodInfo);
                                }
                            }
                        }
                    }

                    return result.toArray(new PaymentMethodInfo[result.size()]);
                }
            }

            return new PaymentMethodInfo[0];
        }
    }

    public static class CheckoutInfo {
        Price price;
        LineItem[] lineItems;
    }

    public static class LineItem {
        String id;
        String refersTo;
        String sku;
        String name;
        String scannedCode;
        int amount;
        int price;
        Integer units;
        int totalPrice;
        LineItemType type;
    }

    public enum LineItemType {
        @SerializedName("default")
        DEFAULT,
        @SerializedName("deposit")
        DEPOSIT,
        @SerializedName("discount")
        DISCOUNT,
        @SerializedName("giveaway")
        GIVEAWAY,
    }

    public static class ExitToken {
        public String value;
        public String format;
    }

    public static class Price {
        int price;
        int netPrice;
    }

    public static class PaymentInformation {
        public String qrCodeContent;
        public String encryptedOrigin;
        public String originType;
        public String validUntil;
        public String cardNumber;
        public String deviceID;
        public String deviceName;
        public String deviceFingerprint;
        public String deviceIPAddress;
    }

    public static class CheckoutProcessRequest {
        public SignedCheckoutInfo signedCheckoutInfo;
        public PaymentMethod paymentMethod;
        public PaymentInformation paymentInformation;
        public String finalizedAt;
        public Boolean processedOffline;
    }

    public static class PaymentMethodInfo {
        public String id;
        public String[] acceptedOriginTypes;
    }

    public static class PaymentResult {
        public String originCandidateLink;
        public String failureCause;
    }

    public enum State {
        @SerializedName("pending")
        PENDING,
        @SerializedName("processing")
        PROCESSING,
        @SerializedName("successful")
        SUCCESSFUL,
        @SerializedName("failed")
        FAILED,
    }

    public enum CheckType {
        @SerializedName("min_age")
        MIN_AGE,
    }

    public enum Performer {
        @SerializedName("app")
        APP,
        @SerializedName("backend")
        SUPERVISOR,
        @SerializedName("supervisor")
        BACKEND,
        @SerializedName("payment")
        PAYMENT
    }

    public static class Check {
        public String id;
        public Map<String, Href> links;
        public CheckType type;
        public Integer requiredAge;
        public Performer performedBy;
        public State state;

        public String getSelfLink() {
            Href link = links.get("self");
            if (link != null && link.href != null) {
                return link.href;
            }
            return null;
        }
    }

    public static class Fulfillment {
        public String id;
        public String type;
        public FulfillmentState state;
        public String[] refersTo;
        public Map<String, Href> links;

        public String getSelfLink() {
            Href link = links.get("self");
            if (link != null && link.href != null) {
                return link.href;
            }
            return null;
        }
    }

    public static class CheckoutProcessResponse {
        public Map<String, Href> links;
        public Boolean supervisorApproval;
        public Boolean paymentApproval;
        public Check[] checks;
        @SerializedName("orderID")
        public String orderId;
        public boolean aborted;
        public JsonObject checkoutInfo;
        public PaymentMethod paymentMethod;
        public boolean modified;
        public PaymentInformation paymentInformation;
        public ExitToken exitToken;
        public State paymentState;
        public PaymentResult paymentResult;
        public Fulfillment[] fulfillments;

        public String getSelfLink() {
            Href link = links.get("self");
            if (link != null && link.href != null) {
                return link.href;
            }
            return null;
        }

        public String getReceiptLink() {
            Href link = links.get("receipt");
            if (link != null && link.href != null) {
                return link.href;
            }
            return null;
        }
    }

    public interface CheckoutInfoResult {
        void success(SignedCheckoutInfo signedCheckoutInfo, int onlinePrice, PaymentMethodInfo[] availablePaymentMethods);
        void noShop();
        void invalidProducts(List<Product> products);
        void noAvailablePaymentMethod();
        void invalidDepositReturnVoucher();
        void unknownError();
        void connectionError();
    }

    public interface PaymentProcessResult {
        void success(CheckoutProcessResponse checkoutProcessResponse, String rawResponse);
        void error();
    }

    public interface PaymentAbortResult {
        void success();
        void error();
    }

    private final Project project;
    private final OkHttpClient okHttpClient;
    private Call call;

    CheckoutApi(Project project) {
        this.project = project;
        this.okHttpClient = project.getOkHttpClient();
    }

    public void cancel() {
        if (call != null) {
            call.cancel();
            call = null;
        }
    }

    public void abort(final CheckoutProcessResponse checkoutProcessResponse, final PaymentAbortResult paymentAbortResult) {
        final Request request = new Request.Builder()
                .url(Snabble.getInstance().absoluteUrl(checkoutProcessResponse.getSelfLink()))
                .patch(RequestBody.create(JSON, "{\"aborted\":true}"))
                .build();

        okHttpClient.newCall(request).enqueue(new Callback() {
            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (response.isSuccessful()) {
                    Logger.d("Payment aborted");
                    if (paymentAbortResult != null) {
                        paymentAbortResult.success();
                    }
                } else {
                    Logger.e("Error while aborting payment");
                    if (paymentAbortResult != null) {
                        paymentAbortResult.error();
                    }
                }
            }

            @Override
            public void onFailure(Call call, IOException e) {
                Logger.e("Error while aborting payment");
                if (paymentAbortResult != null) {
                    paymentAbortResult.error();
                }
            }
        });
    }

    public void createCheckoutInfo(final ShoppingCart.BackendCart backendCart,
                                   final PaymentMethod[] clientAcceptedPaymentMethods,
                                   final CheckoutInfoResult checkoutInfoResult,
                                   final long timeout) {
        String checkoutUrl = project.getCheckoutUrl();
        if (checkoutUrl == null) {
            Logger.e("Could not checkout, no checkout url provided in metadata");
            checkoutInfoResult.connectionError();
            return;
        }

        final Request request = new Request.Builder()
                .url(Snabble.getInstance().absoluteUrl(checkoutUrl))
                .post(RequestBody.create(JSON, GsonHolder.get().toJson(backendCart)))
                .build();

        cancel();

        OkHttpClient okClient = okHttpClient;

        if (timeout > 0) {
            okClient = okClient.newBuilder()
                    .callTimeout(timeout, TimeUnit.MILLISECONDS)
                    .build();
        }

        call = okClient.newCall(request);
        call.enqueue(new JsonCallback<SignedCheckoutInfo, JsonObject>(SignedCheckoutInfo.class, JsonObject.class) {
            @Override
            public void success(SignedCheckoutInfo signedCheckoutInfo) {
                int price;

                if (signedCheckoutInfo.checkoutInfo.has("price")
                        && signedCheckoutInfo.checkoutInfo.get("price").getAsJsonObject().has("price")) {
                    price = signedCheckoutInfo.checkoutInfo
                            .get("price")
                            .getAsJsonObject()
                            .get("price")
                            .getAsInt();
                } else {
                    price = project.getShoppingCart().getTotalPrice();
                }

                PaymentMethodInfo[] availablePaymentMethods = signedCheckoutInfo.getAvailablePaymentMethods(clientAcceptedPaymentMethods);
                if (availablePaymentMethods != null && availablePaymentMethods.length > 0) {
                    checkoutInfoResult.success(signedCheckoutInfo, price, availablePaymentMethods);
                } else {
                    checkoutInfoResult.connectionError();
                }
            }

            @Override
            public void failure(JsonObject obj) {
                try {
                    JsonObject error = obj.get("error").getAsJsonObject();
                    String type = error.get("type").getAsString();

                    switch (type) {
                        case "invalid_cart_item":
                            List<String> invalidSkus = new ArrayList<>();
                            JsonArray arr = error.get("details").getAsJsonArray();

                            for (int i=0; i<arr.size(); i++) {
                                String sku = arr.get(0).getAsJsonObject().get("sku").getAsString();
                                invalidSkus.add(sku);
                            }

                            List<Product> invalidProducts = new ArrayList<>();
                            ShoppingCart cart = project.getShoppingCart();
                            for (int i=0; i<cart.size(); i++) {
                                Product product = cart.get(i).getProduct();
                                if (product != null) {
                                    if (invalidSkus.contains(product.getSku())) {
                                        invalidProducts.add(product);
                                    }
                                }
                            }

                            Logger.e("Invalid products");
                            checkoutInfoResult.invalidProducts(invalidProducts);
                            break;
                        case "no_available_method":
                            checkoutInfoResult.noAvailablePaymentMethod();
                            break;
                        case "bad_shop_id":
                        case "shop_not_found":
                            checkoutInfoResult.noShop();
                            break;
                        case "invalid_deposit_return_voucher":
                            checkoutInfoResult.invalidDepositReturnVoucher();
                            break;
                        default:
                            checkoutInfoResult.unknownError();
                            break;
                    }
                } catch (Exception e) {
                    error(e);
                }
            }

            @Override
            public void error(Throwable t) {
                Logger.e("Error creating checkout info: " + t.getMessage());
                checkoutInfoResult.connectionError();
            }
        });
    }

    public void updatePaymentProcess(final String url,
                                     final PaymentProcessResult paymentProcessResult) {
        final Request request = new Request.Builder()
                .url(Snabble.getInstance().absoluteUrl(url))
                .get()
                .build();

        cancel();

        call = okHttpClient.newCall(request);
        call.enqueue(new SimpleJsonCallback<CheckoutProcessResponse>(CheckoutProcessResponse.class) {
            @Override
            public void success(CheckoutProcessResponse checkoutProcessResponse) {
                paymentProcessResult.success(checkoutProcessResponse, rawResponse());
                CheckoutApi.this.call = null;
            }

            @Override
            public void error(Throwable t) {
                paymentProcessResult.error();
            }
        });
    }

    public void updatePaymentProcess(final CheckoutProcessResponse checkoutProcessResponse,
                                     final PaymentProcessResult paymentProcessResult) {
        String url = checkoutProcessResponse.getSelfLink();
        if (url == null) {
            paymentProcessResult.error();
            return;
        }

        updatePaymentProcess(url, paymentProcessResult);
    }

    public void createPaymentProcess(final String id,
                                     final SignedCheckoutInfo signedCheckoutInfo,
                                     final PaymentMethod paymentMethod,
                                     final PaymentCredentials paymentCredentials,
                                     final boolean processedOffline,
                                     final Date finalizedAt,
                                     final PaymentProcessResult paymentProcessResult) {
        CheckoutProcessRequest checkoutProcessRequest = new CheckoutProcessRequest();
        checkoutProcessRequest.paymentMethod = paymentMethod;
        checkoutProcessRequest.signedCheckoutInfo = signedCheckoutInfo;

        if (processedOffline) {
            checkoutProcessRequest.processedOffline = true;

            if (finalizedAt != null) {
                checkoutProcessRequest.finalizedAt = DateUtils.toRFC3339(finalizedAt);
            }
        }

        if (paymentCredentials != null) {
            checkoutProcessRequest.paymentInformation = new PaymentInformation();

            String data = paymentCredentials.getEncryptedData();
            if (data == null) {
                paymentProcessResult.error();
                return;
            }

            checkoutProcessRequest.paymentInformation.originType = paymentCredentials.getType().getOriginType();
            checkoutProcessRequest.paymentInformation.encryptedOrigin = data;

            if (paymentCredentials.getType() == PaymentCredentials.Type.CREDIT_CARD_PSD2) {
                Date date = new Date(paymentCredentials.getValidTo());
                SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy/MM/dd");
                checkoutProcessRequest.paymentInformation.validUntil = simpleDateFormat.format(date);
                checkoutProcessRequest.paymentInformation.cardNumber = paymentCredentials.getObfuscatedId();
            } else if (paymentCredentials.getType() == PaymentCredentials.Type.PAYDIREKT) {
                Map<String, String> additionalData = paymentCredentials.getAdditionalData();
                if (additionalData != null) {
                    checkoutProcessRequest.paymentInformation.deviceID = additionalData.get("deviceID");
                    checkoutProcessRequest.paymentInformation.deviceName = additionalData.get("deviceName");
                    checkoutProcessRequest.paymentInformation.deviceFingerprint = additionalData.get("deviceFingerprint");
                    checkoutProcessRequest.paymentInformation.deviceIPAddress = additionalData.get("deviceIPAddress");
                }
            }
        }

        String url = signedCheckoutInfo.getCheckoutProcessLink();
        if (url == null) {
            paymentProcessResult.error();
            return;
        }

        url = url + "/" + id;
        String json = GsonHolder.get().toJson(checkoutProcessRequest);
        final Request request = new Request.Builder()
                .url(Snabble.getInstance().absoluteUrl(url))
                .put(RequestBody.create(JSON, json))
                .build();

        cancel();

        call = okHttpClient.newCall(request);

        String finalUrl = url;
        call.enqueue(new SimpleJsonCallback<CheckoutProcessResponse>(CheckoutProcessResponse.class) {
            @Override
            public void success(CheckoutProcessResponse checkoutProcess) {
                paymentProcessResult.success(checkoutProcess, rawResponse());
            }

            @Override
            public void error(Throwable t) {
                if (responseCode() == 403) {
                    updatePaymentProcess(finalUrl, paymentProcessResult);
                } else {
                    paymentProcessResult.error();
                }
            }
        });
    }
}

File: core/src/main/java/io/snabble/sdk/checkout/CheckoutApi.kt
package io.snabble.sdk.checkout

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.annotations.SerializedName
import com.google.gson.reflect.TypeToken
import io.snabble.sdk.FulfillmentState
import io.snabble.sdk.PaymentMethod
import io.snabble.sdk.Product
import io.snabble.sdk.coupons.Coupon
import io.snabble.sdk.payment.PaymentCredentials
import io.snabble.sdk.shoppingcart.data.cart.BackendCart
import java.io.Serializable
import java.util.Date

/**
 * Interface for the snabble Checkout API
 *
 * Backend API Documentation:
 * https://docs.snabble.io/docs/api/api_checkout
 */
interface CheckoutApi {
    /**
     * Cancel all operations
     */
    fun cancel()

    fun abort(
        checkoutProcessResponse: CheckoutProcessResponse,
        paymentAbortResult: PaymentAbortResult?
    )

    /**
     * Creates a checkout info with mandatory price calculation and available payment methods.
     * This document can be used to show the real price to the user and it can be used
     * to start a checkout process as input of {@link #createPaymentProcess}
     */
    fun createCheckoutInfo(
        backendCart: BackendCart,
        checkoutInfoResult: CheckoutInfoResult? = null,
        timeout: Long = -1
    )

    /**
     *  Updates an existing payment process
     */
    fun updatePaymentProcess(
        checkoutProcessResponse: CheckoutProcessResponse,
        paymentProcessResult: PaymentProcessResult?
    )

    /**
     * Creates a payment process using stored payment credentials and a signed checkout info from
     * create checkout info.
     */
    fun createPaymentProcess(
        id: String,
        signedCheckoutInfo: SignedCheckoutInfo,
        paymentMethod: PaymentMethod,
        processedOffline: Boolean,
        paymentCredentials: PaymentCredentials?,
        finalizedAt: Date?,
        paymentProcessResult: PaymentProcessResult?
    )

    /**
     * Authorize a payment of a existing checkout process providing payment credentials.
     *
     * Only used for one-time token payments like google pay.
     */
    fun authorizePayment(
        checkoutProcessResponse: CheckoutProcessResponse,
        authorizePaymentRequest: AuthorizePaymentRequest,
        authorizePaymentResult: AuthorizePaymentResult?
    )
}

interface AuthorizePaymentResult {
    fun onSuccess()
    fun onError()
}

interface CheckoutInfoResult {
    fun onSuccess(
        signedCheckoutInfo: SignedCheckoutInfo,
        onlinePrice: Int,
        availablePaymentMethods: List<PaymentMethodInfo>
    )

    fun onNoShopFound()
    fun onInvalidProducts(products: List<Product>)
    fun onInvalidItems(itemIds: List<String>)
    fun onNoAvailablePaymentMethodFound()
    fun onUnknownError()
    fun onConnectionError()
}

interface PaymentProcessResult {
    fun onSuccess(checkoutProcessResponse: CheckoutProcessResponse?, rawResponse: String?)
    fun onError()
    fun onNotFound()
}

interface PaymentAbortResult {
    fun onSuccess()
    fun onError()
}

enum class LineItemType {
    @SerializedName("default") DEFAULT,
    @SerializedName("deposit") DEPOSIT,
    @SerializedName("discount") DISCOUNT,
    @SerializedName("coupon") COUPON,
    @SerializedName("depositReturnVoucher") DEPOSIT_RETURN_VOUCHER,
    @SerializedName("depositReturn") DEPOSIT_RETURN
}

enum class CheckState {
    @SerializedName("unauthorized") UNAUTHORIZED,
    @SerializedName("pending") PENDING,
    @SerializedName("processing")  PROCESSING,
    @SerializedName("authenticating") AUTHENTICATING,
    @SerializedName("successful") SUCCESSFUL,
    @SerializedName("transferred") TRANSFERRED,
    @SerializedName("failed") FAILED
}

enum class CheckType {
    @SerializedName("min_age") MIN_AGE,
    @SerializedName("supervisor_approval") SUPERVISOR
}

enum class Performer {
    @SerializedName("app") APP,
    @SerializedName("supervisor") SUPERVISOR,
    @SerializedName("backend") BACKEND,
    @SerializedName("payment") PAYMENT
}

enum class RoutingTarget {
    @SerializedName("gatekeeper") GATEKEEPER,
    @SerializedName("supervisor") SUPERVISOR,
    @SerializedName("none") NONE
}

data class Href(
    @SerializedName("href") val href: String? = null,
) : Serializable

data class SignedCheckoutInfo(
    @SerializedName("checkoutInfo") val checkoutInfo: JsonObject? = null,
    @SerializedName("signature") val signature: String? = null,
    @SerializedName("links") val links: Map<String, Href>? = null,
) {
    val checkoutProcessLink: String?
        get() = links?.get("checkoutProcess")?.href

    val isRequiringTaxation: Boolean
        get() {
            try {
                if (checkoutInfo?.has("requiredInformation") == true) {
                    val jsonArray = checkoutInfo.getAsJsonArray("requiredInformation")
                    jsonArray?.forEach { element ->
                        val id = element.asJsonObject["id"].asString
                        val hasValue = element.asJsonObject.has("value")
                        if (id == "taxation" && !hasValue) {
                            return true
                        }
                    }
                }
            } catch (e: Exception) {
                return false
            }
            return false
        }

    fun getAvailablePaymentMethods(): List<PaymentMethodInfo> {
        if (checkoutInfo?.has("paymentMethods") == true) {
            val jsonArray = checkoutInfo.getAsJsonArray("paymentMethods")
            if (jsonArray != null) {
                val type = object : TypeToken<List<PaymentMethodInfo>?>() {}.type
                return Gson().fromJson(jsonArray, type)
            }
        }
        return emptyList()
    }

    fun getRedeemedCoupons(availableCoupons: List<Coupon>): List<Coupon> {
        val redeemedCoupons = mutableListOf<Coupon>()
        checkoutInfo?.get("lineItems")?.let {
            it.asJsonArray?.let { jsonArray ->
                val lineItems = Gson().fromJson<List<LineItem>>(
                    jsonArray,
                    object : TypeToken<List<LineItem?>?>() {}.type
                )
                lineItems.forEach { lineItem ->
                    if (lineItem.type == LineItemType.COUPON && lineItem.redeemed) {
                        availableCoupons.forEach { coupon ->
                            if (coupon.id == lineItem.couponId) {
                                redeemedCoupons.add(coupon)
                            }
                        }
                    }
                }
            }
        }
        return redeemedCoupons
    }
}

data class CheckoutInfo(
    @SerializedName("price") val price: Price? = null,
    @SerializedName("lineItems") val lineItems: List<LineItem> = emptyList(),
    @SerializedName("violations") val violations: List<Violation> = emptyList(),
)

data class Violation(
    @SerializedName("type") val type: ViolationType? = null,
    @SerializedName("refersTo") val refersTo: String? = null,
    @SerializedName("message") val message: String? = null,
)

enum class ViolationType {
    @SerializedName("deposit_return_voucher_already_redeemed")
    DEPOSIT_RETURN_ALREADY_REDEEMED,

    @SerializedName("deposit_return_voucher_duplicate")
    DEPOSIT_RETURN_DUPLICATED,

    @SerializedName("coupon_already_voided")
    COUPON_ALREADY_VOIDED,

    @SerializedName("coupon_currently_not_valid")
    COUPON_CURRENTLY_NOT_VALID,

    @SerializedName("coupon_invalid")
    COUPON_INVALID,
}

data class LineItem(
    @SerializedName("id") val id: String? = null,
    @SerializedName("amount") val amount: Int = 0,
    @SerializedName("couponID") val couponId: String? = null,
    @SerializedName("discountID") val discountID: String? = null,
    @SerializedName("discountRuleID") val discountRuleID: String? = null,
    @SerializedName("discountType") val discountType: String? = null,
    @SerializedName("listPrice") val listPrice: Int = 0,
    @SerializedName("name") val name: String? = null,
    @SerializedName("price") val price: Int = 0,
    @SerializedName("priceModifiers") val priceModifiers: List<PriceModifier>? = null,
    @SerializedName("redeemed") val redeemed: Boolean = false,
    @SerializedName("refersTo") val refersTo: String? = null,
    @SerializedName("referenceUnit") val referenceUnit: String? = null,
    @SerializedName("scannedCode") val scannedCode: String? = null,
    @SerializedName("sku") val sku: String? = null,
    @SerializedName("totalPrice") val totalPrice: Int = 0,
    @SerializedName("type") val type: LineItemType? = null,
    @SerializedName("units") val units: Int? = null,
    @SerializedName("weight") val weight: Int? = null,
    @SerializedName("weightUnit") val weightUnit: String? = null,
)

data class PriceModifier(
    @SerializedName("name") val name: String? = null,
    @SerializedName("price") val price: Int = 0,
    @SerializedName("action") val action: String? = null
)

data class ExitToken(
    @SerializedName("value") val value: String? = null,
    @SerializedName("format") val format: String? = null,
)

data class Price(
    @SerializedName("price") val price: Int = 0,
    @SerializedName("netPrice") val netPrice: Int = 0,
)

data class PaymentInformation(
    @SerializedName("qrCodeContent") val qrCodeContent: String? = null,
    @SerializedName("encryptedOrigin") val encryptedOrigin: String? = null,
    @SerializedName("originType") val originType: String? = null,
    @SerializedName("validUntil") val validUntil: String? = null,
    @SerializedName("cardNumber") val cardNumber: String? = null,
    @SerializedName("deviceID") val deviceID: String? = null,
    @SerializedName("deviceName") val deviceName: String? = null,
    @SerializedName("deviceFingerprint") val deviceFingerprint: String? = null,
    @SerializedName("deviceIPAddress") val deviceIPAddress: String? = null,
    @SerializedName("handoverInformation") val handoverInformation: String? = null,
    @SerializedName("subject") val subject: String? = null
)

data class CheckoutProcessRequest(
    @SerializedName("signedCheckoutInfo") val signedCheckoutInfo: SignedCheckoutInfo? = null,
    @SerializedName("paymentMethod") val paymentMethod: PaymentMethod? = null,
    @SerializedName("paymentInformation") val paymentInformation: PaymentInformation? = null,
    @SerializedName("finalizedAt") val finalizedAt: String? = null,
    @SerializedName("processedOffline") val processedOffline: Boolean? = null,
)

data class PaymentMethodInfo(
    @SerializedName("id") val id: String? = null,
    @SerializedName("isTesting") val isTesting: Boolean = false,
    @SerializedName("acceptedOriginTypes") val acceptedOriginTypes: List<String> = emptyList()
)

data class PaymentResult(
    @SerializedName("originCandidateLink") val originCandidateLink: String? = null,
    @SerializedName("failureCause") val failureCause: String? = null,
)

data class AuthorizePaymentRequest(
    @SerializedName("encryptedOrigin") val encryptedOrigin: String? = null,
    @SerializedName("name") val name: String? = null,
    @SerializedName("countryCode") val countryCode: String? = null,
    @SerializedName("state") val state: String? = null
)

data class Check(
    @SerializedName("id") val id: String? = null,
    @SerializedName("links") val links: Map<String, Href>? = null,
    @SerializedName("type") val type: CheckType? = null,
    @SerializedName("requiredAge") val requiredAge: Int? = null,
    @SerializedName("performedBy") val performedBy: Performer? = null,
    @SerializedName("state") val state: CheckState? = null,
) {
    val selfLink: String?
        get() = links?.get("self")?.href
}

data class Fulfillment(
    @SerializedName("id") val id: String? = null,
    @SerializedName("type") val type: String? = null,
    @SerializedName("state") val state: FulfillmentState? = null,
    @SerializedName("refersTo") val refersTo: List<String> = emptyList(),
    @SerializedName("links") val links: Map<String, Href>? = null,
) {
    val selfLink: String?
        get() = links?.get("self")?.href
}

data class DepositReturnVoucher(
    @SerializedName("refersTo")
    val refersTo: String,
    @SerializedName("state")
    val state: DepositReturnVoucherState
)

enum class DepositReturnVoucherState {
    @SerializedName("pending")
    PENDING,

    @SerializedName("redeemed")
    REDEEMED,

    @SerializedName("redeemingFailed")
    REDEEMING_FAILED,

    @SerializedName("rolledback")
    ROLLED_BACK,

    @SerializedName("rollbackFailed")
    ROLLBACK_FAILED
}

data class CheckoutProcessResponse(
    @SerializedName("links") val links: Map<String, Href>? = null,
    @SerializedName("checks") val checks: List<Check> = emptyList(),
    @SerializedName("orderID") val orderId: String? = null,
    @SerializedName("depositReturnVouchers") val depositReturnVouchers: List<DepositReturnVoucher>? = null,
    @SerializedName("aborted") val aborted: Boolean = false,
    @SerializedName("paymentMethod") val paymentMethod: PaymentMethod? = null,
    @SerializedName("paymentInformation") val paymentInformation: PaymentInformation? = null,
    @SerializedName("paymentPreauthInformation") val paymentPreauthInformation: JsonObject? = null,
    @SerializedName("exitToken") val exitToken: ExitToken? = null,
    @SerializedName("paymentState") val paymentState: CheckState? = null,
    @SerializedName("pricing") val pricing: Pricing? = null,
    @SerializedName("routingTarget") val routingTarget: RoutingTarget? = null,
    @SerializedName("paymentResult") val paymentResult: PaymentResult? = null,
    @SerializedName("fulfillments") val fulfillments: List<Fulfillment> = emptyList(),
) {
    val selfLink: String?
        get() = links?.get("self")?.href

    val authorizePaymentLink: String?
        get() = links?.get("authorizePayment")?.href

    val paymentRedirect: String?
        get() = links?.get("paymentRedirect")?.href

    val originCandidateLink: String?
        get() = paymentResult?.originCandidateLink
}

data class Pricing(
    @SerializedName("price") val price: Price? = null,
)
